### ISSUE TITLE
DefaultHttp2ConnectionEncoder.writePriority is broken.

### ISSUE THREAD

**Issue by buchgr** [netty/netty#3517]
If trying to change the priority of a stream that has been closed and removed ("http2 gced") already, the `DefaultHttp2ConnectionEncoder.writePriority(...)` method will attempt to create a new stream and fail, because the stream id will be smaller than the next expected stream id.

I don't have the time right now to think of a good way to fix this, as I also expect that this might have to come with some bigger changes.

**Comment by Scottmitch** [comment]
@buchgr - Thanks for reporting! Care to elaborate on the bigger changes you anticipate?  This may depend on the resolution of https://github.com/netty/netty/issues/3518 but I'm hoping we can just write the frame and not create the stream (not track in priority tree).

**Comment by nmittler** [comment]
@Scottmitch @buchgr some thoughts on this...

If the stream does not exist but MAY have existed, we could just ignore the parent altogether and re-prioritize the stream such that it has no parent.

Otherwise, we try to create the stream. If creating the parent fails in this case, it's a legitimate failure since we're trying to create a new stream that has not yet existed.

WDYT?

**Comment by Scottmitch** [comment]
@nmittler - Can you clarify what ignoring the parent buys us and where that leaves us?  Seems like this would change the semantics of the method call.  I agree that for streams that MAY have previously existed (based upon `nextStreamId()` and `createdStreamId(...)`) we should handle this (at least not throw an error), and otherwise just propagate the error.  I'm trying to think of a way to avoid creating the streams, and honoring the semantics of the request, but I haven't come up with anything.  Here are my concerns for re-creating the streams:

1. Clean up condition.  The resolution for issue https://github.com/netty/netty/issues/3518 is relatively self contained assuming streams will eventually transition to closed, and the their subtrees will be analyzed for cleaned up.  Re-inserting streams into the tree muddies the waters a bit....the closed stream would be violating these assumptions as soon as they are inserted. Also these "zombie streams" may be inadvertently cleaned up if they are in a subtree that just so happens to transition to all closed.
2. Memory usage.  Should we be circumventing `maxActiveStreams` when creating streams for this case?  I think not...but the streams are not active...soo?
3. Both streams must be considered.  We have primarily discussed the stream we are writing the frame for but we have not discussed the dependency (aka parent) stream.  What if the parent stream is closed (and we can't get a stream object for it), then we will be in the same situation for the parent stream.

Another part to this issue is the receive side.  We are throwing a connection error if we receive a PRIORITY frame...we should not do this.  We should either:

1. Do something similar to the send side; OR
2. Ignore it [section 5.1](https://tools.ietf.org/html/draft-ietf-httpbis-http2-17#section-5.1):

>  PRIORITY frames can be sent on closed streams to prioritize
      streams that are dependent on the closed stream.  Endpoints SHOULD
      process PRIORITY frames, though they can be ignored if the stream
      has been removed from the dependency tree (see Section 5.3.4).

**Comment by nmittler** [comment]
@Scottmitch sorry I think I must have poorly communicated the thought, let me try again ...

Sending PRIORITY for a removed stream:
I wasn't suggesting that we create/add a new stream to simulate the removed stream.  Now that I think about it, this case can probably be simplified to essentially letting the receiving-side deal with it.  On the sending side, we should only create a new stream if it could not have previously existed, otherwise we just send the PRIORITY frame.

Receiving PRIORITY for a removed stream:
I think ignoring makes sense here as long as the stream MAY have existed.

WDYT?

**Comment by Scottmitch** [comment]
@nmittler - Yes this is what I was originally hoping for :)  I don't want to get into the re-creating streams business.

**Comment by nmittler** [comment]
@Scottmitch is this resolved by #3544?

**Comment by Scottmitch** [comment]
@nmittler +1.  Thanks!

**Comment by buchgr** [comment]
@Scottmitch thanks!
